packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -30,6 +30,7 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
   const rowsLookup = useGridSelector(apiRef, gridRowsLookupSelector);
   const isMultipleKeyPressed = useGridSelector(apiRef, gridKeyboardMultipleKeySelector);
   const allowMultipleSelectionKeyPressed = React.useRef<boolean>(false);
+  const isMultipleKeyPressedRef = React.useRef<boolean>(false);
```

Comment:
We shouldn't need for this. Isn't it `selectRowModel` that is called wight the wrong argument? 

```suggestion
```

Reply:
I think it is called with the right argument. 
https://github.com/mui-org/material-ui-x/blob/c992eb3de48ed3f465cb8d3daacd45f13758224a/packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts#L166-L173

For deselection, key press should be detected. Just clicking on a selected row should not be deselected, but with a combination of supported keypress + click.
Not sure about the consideration of ref though. I considered it so that the key press boolean value can be persisted to check its value below.

Reply:
Can't we just use `isMultipleKeyPressed` directly?

Reply:
> Can't we just use `isMultipleKeyPressed` directly?

No. I think when we click on the row the DataGrid re-renders and `isMultipleKeyPressed ` is initialised back to false. Storing it in a ref means the value is persisted by React across re-renders. Also, it's value change to `false` (if we leave the pressing of key) will not cause a re-render.

Reply:
This sounds like a great opportunity to refactor the logic. I would propose we remove `isMultipleKeyPressed` altogether. It's the last location we haven't yet refactored to remove the concept. multi-selection has the same bugs we had with multi-sorting. e.g. if you press <kbd>CTRL</kbd> outside of the data grid, it fails: #1204

Reply:
To drop `isMultipleKeyPressed `, you just have to check if a multiple key is pressed in `handleRowClick` and set the third argument of `selectRowModel` to true. Don't forget to also check if the `disableMultipleSelection` prop is not true. I did something similar to fix the multi-sorting:

https://github.com/mui-org/material-ui-x/blob/c992eb3de48ed3f465cb8d3daacd45f13758224a/packages/grid/_modules_/grid/hooks/features/sorting/useGridSorting.ts#L241

Reply:
Thanks, I'll refactor to drop `isMultipleKeyPressed`. 

> Don't forget to also check if the disableMultipleSelection prop is not true

I think you mean check for `disableMultipleSelection` to be true for DataGrid's single selection?

Reply:
I've refactored in https://github.com/mui-org/material-ui-x/pull/1813/commits/4021a0f2aeb8d8246cb633a87f44b2b668409c16